uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/IFirstPartyApp.java

```diff
@@ -0,0 +1,71 @@
+//  Copyright (c) Microsoft Corporation.
+//  All rights reserved.
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.client.ui.automation.app;
+
+import androidx.annotation.NonNull;
+
+import com.microsoft.identity.client.ui.automation.interaction.FirstPartyAppPromptHandlerParameters;
+
+/**
+ * An interface describing a first party application and the actions that can be performed on them.
+ */
+public interface IFirstPartyApp {
+
+    /**
+     * Add the first user account to this first party app
+     *
+     * @param username                the username of the account to add
+     * @param password                the password of the account to add
+     * @param promptHandlerParameters the prompt handler parameters indicating how to handle prompt
+     */
+    void addFirstAccount(@NonNull final String username,
+                         @NonNull final String password,
+                         @NonNull final FirstPartyAppPromptHandlerParameters promptHandlerParameters);
+
+    /**
+     * Add another account to this first party app. This must only be called if the an account was
+     * previously added to this first party app.
+     *
+     * @param username                the username of the account to add
+     * @param password                the password of the account to add
+     * @param promptHandlerParameters the prompt handler parameters indicating how to handle prompt
+     */
+    void addAnotherAccount(final String username,
```

Comment:
I like your suggestion....we can track of how many accounts are added internally in our app model classes and just create one method here `addAccount` and change the behavior of it depending on if there was already an account added or not. I think this make sense and I'll incorporate that change.

Reply:
This actually isn't that simple because some apps like Microsoft Word, simply launching them adds an account to the app (without user interaction). So a caller could simply call `word.launch()` and that would add the first account to the app.

I think we can also make improvements here overall in terms of when someone calls `addAccount` then is it their repsonsilbility to make sure that they have called `launch` prior to that. The code currently assumes that the app is launched and open when someone calls `addAccount` and will fail the test case if the app isn't open because it wouldn't find the appropriate UI element. Looking more into UI Automator, I think I can use the [`getCurrentPackageName`](https://developer.android.com/reference/androidx/test/uiautomator/UiDevice#getCurrentPackageName()) internally in `addAccount` to decide if we also need to launch the app or not. I'm thinking I'll leave the code as is for now, and create a backlog item for these changes and address both of these (distinct methods for adding accounts, and us vs them launching the app) next month.

Reply:
@AdamBJohnsonx ^^

Reply:
Ew.  Sounds like there's a plan, then.